### ISSUE TITLE
warn_cstruct_memaccess should not trigger inside constructors for otherwise trivially initializable types

### ISSUE THREAD

**Issue by ojhunt** [llvm/llvm-project#156996]
We warn on this code, and variations of it, when gcc does not

```cpp
  struct Object {
   int some_field;
   Object() { memset(this, 0, sizeof(*this); }
   Object(const Object&);
  };
```

I put together a general set of the cases at https://godbolt.org/z/5YYj1oxf9

Basically performing these operations within constructors of a type, as long as the reason for the type not being `trivially_{whatever}` is just its own constructors, not due to the object being polymorphic or some sub- or baseobject being non-trivial to copy or initialize.

**Comment by ojhunt** [comment]
These could be related to #111434 @serge-sans-paille, but @ahatanak has also written some of these checks.

**Comment by ojhunt** [comment]
This makes me wonder if we should warn on the use of the normal memset/bzero functions in destructors on the basis that they're likely being used to try an "securely" clear an object, but it would be UB for that to be observable so can be (is?) removed by the compiler. similarly `memset/bzero(ptr, ...); delete ptr/free(ptr);`
